### PR TITLE
[SPARK-32268][SQL] Dynamic bloom filter join pruning

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

Reduce the shuffle data can significantly improve the query performance and increase Spark cluster stability.

This PR implements dynamic bloom filter join pruning to reduce the shuffle data. The main changes:
- Add `BuildBloomFilter` and `InBloomFilter` UDF.
- Enhance `RepartitionByExpression` to support shuffle with `ENSURE_REQUIREMENTS` origin. Dynamic bloom filter join pruning use it to reuse exchange.
- Add a new rule `DynamicBloomFilterPruning`.
- Dynamic bloom filter join pruning supports AQE.
- Support dynamic bloom filter nested with dynamic partition pruning when AQE enabled.
- Add an AQE optimization rule(`OptimizeBloomFilterJoin`) to update the `expectedNumItems` base on stats collected by AQE which is more accurate.

### Why are the changes needed?

Improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test and benchmark test:

SQL | Before this PR | After this PR
--- | --- | ---
q12 | 11 | 8
q20 | 6 | 6
q37 | 29 | 13
q50 | 78 | 29
q80 | 14 | 14
q82 | 38 | 16
q93 | 126 | 108
q98 | 6 | 6

</body>

</html>

### CODE REVIEW

**Comment by wangyum** [comment]
Example of TPC-DS q37.

Before this PR | After this PR
--- | ---
image | image

**Comment by HyukjinKwon** [comment]
I think we should probably at least have a design doc to explain this .. from a cursory look the change looks huge.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -353,6 +353,23 @@ object SQLConf {
       .checkValue(ratio => ratio > 0.0 && ratio <= 1.0, "The ratio value must be in (0.0, 1.0].")
       .createWithDefault(0.04)
 
+  val DYNAMIC_BLOOM_FILTER_JOIN_PRUNING_ENABLED =
+    buildConf("spark.sql.optimizer.dynamicBloomFilterJoinPruning.enabled")
+      .doc("When true, we will generate bloom filter predicate for column when it's used as join " +
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:358]: nit: "generates a bloom filter predicate for a join key column"

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -353,6 +353,23 @@ object SQLConf {
...
+        "key. Note that, dynamic bloom filter join pruning only works if exchange reuse enabled.")
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:359]: Use the config property so it becomes `s"only works with ${SQLConf.EXCHANGE_REUSE_ENABLED.key} enabled"`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -353,6 +353,23 @@ object SQLConf {
...
+      .version("3.3.0")
+      .booleanConf
+      .createWithDefault(true)
+
+  val DYNAMIC_BLOOM_FILTER_JOIN_PRUNING_MAX_BLOOM_FILTER_ENTRIES =
+    buildConf("spark.sql.optimizer.dynamicBloomFilterJoinPruning.maxBloomFilterEntries")
+      .doc("The maximum number of bloom filter entries allowed when building dynamic bloom filter" +
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:366]: nit: s/when building/for

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:358]: Thank you @jaceklaskowski

**Comment by wangyum** [comment]
> I think we should probably at least have a design doc to explain this .. from a cursory look the change looks huge.

I’m still doing an evaluation in our production environment, and I’ll prepare the design doc after the evaluation done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:359]: Fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 3 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:366]: Fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/BuildBloomFilter.scala [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+  override def createAggregationBuffer(): BloomFilter = {
+    BloomFilter.create(expectedNumItems, expectedNumItems.toDouble / 3000000000L.toDouble)
+  }
+
+  override def update(buffer: BloomFilter, input: InternalRow): BloomFilter = {
+    val value = child.eval(input)
+    // Ignore empty rows
+    if (value != null) {
+      child.dataType match {
```
Comment (zhengruifeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/BuildBloomFilter.scala:53]: we can use a hashed value `Murmur3Hash/XxHash64(children, 42)` instead, then we do not need to check the datatype and can support multi-column with single bloomfilter.

**Pull request closed without merging**